### PR TITLE
More informative Error message on Vulkan driver crash

### PR DESCRIPTION
**Pull request by Megamouse**
The message might not be the only reason, but at least it might help someone like me, who had no idea what he was looking at.

### CODE REVIEW

File: rpcs3/Emu/RSX/VK/vkutils/shared.h [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 #include "../VulkanAPI.h"
+#include <string>
 
 namespace vk
 {
 #define CHECK_RESULT(expr) { VkResult _res = (expr); if (_res != VK_SUCCESS) vk::die_with_error(_res); }
+#define CHECK_RESULT_EXT(expr, msg) { VkResult _res = (expr); if (_res != VK_SUCCESS) vk::die_with_error(_res, msg); }
 
-	void die_with_error(VkResult error_code,
+	void die_with_error(VkResult error_code, std::string message = {},
```
Comment (sampletext32) [inline rpcs3/Emu/RSX/VK/vkutils/shared.h:11]: I wonder if message can be a reference

Comment (Megamouse) [reply rpcs3/Emu/RSX/VK/vkutils/shared.h:11]: does it really matter if we are killing the whole program ?

Comment (sampletext32) [reply rpcs3/Emu/RSX/VK/vkutils/shared.h:11]: sure, it's nothing, just pretending i'm a pro

Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/shared.h:11]: Use const ref anyway, its just nicer.
const std::string& message = {}

File: rpcs3/Emu/RSX/VK/vkutils/device.cpp [hunk 2]
```diff
@@ -356,7 +358,7 @@ namespace vk
 			rsx_log.notice("GPU/driver lacks support for float16 data types. All float16_t arithmetic will be emulated with float32_t.");
 		}
 
-		CHECK_RESULT(vkCreateDevice(*pgpu, &device, nullptr, &dev));
+		CHECK_RESULT_EXT(vkCreateDevice(*pgpu, &device, nullptr, &dev), message_on_error);
```
Comment (kd-11) [inline rpcs3/Emu/RSX/VK/vkutils/device.cpp:361]: CHECK_RESULT_EX sounds better and is consistent with win32 extended API naming which is more familiar to people like me.

File: rpcs3/Emu/RSX/VK/vkutils/shared.cpp [hunk 3]
```diff
@@ -99,7 +99,8 @@ namespace vk
 		{
 		default:
 		case 0:
-			fmt::throw_exception("Assertion Failed! Vulkan API call failed with unrecoverable error: %s%s", error_message, src_loc{line, col, file, func});
+			if (!message.empty()) message += "\n\n";
+			fmt::throw_exception("%sAssertion Failed! Vulkan API call failed with unrecoverable error: %s%s", message, error_message, src_loc{line, col, file, func});
```
Comment (kd-11) [inline rpcs3/Emu/RSX/VK/vkutils/shared.cpp:103]: Ftr, for some people like myself the fmt error is printed on the commandline and never shows a messagebox. This is going to be doubly true when the exception is raised in the routines that initialize the game window for some reason, but I may be the exception, not the rule. Still something to keep in mind.

File: rpcs3/Emu/RSX/VK/vkutils/shared.h (hunk 1 above)
Comment (Megamouse) [reply rpcs3/Emu/RSX/VK/vkutils/shared.h:11]: But I modify the message if not empty

File: rpcs3/Emu/RSX/VK/vkutils/device.cpp (hunk 2 above)
Comment (Megamouse) [reply rpcs3/Emu/RSX/VK/vkutils/device.cpp]: fixed

File: rpcs3/Emu/RSX/VK/vkutils/shared.cpp (hunk 3 above)
Comment (Megamouse) [reply rpcs3/Emu/RSX/VK/vkutils/shared.cpp:103]: Maybe try to find out why it doesn't work for you then.
Should be easy enough to debug report_fatal_error function.

File: rpcs3/Emu/RSX/VK/vkutils/shared.h (hunk 1 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/shared.h:11]: I'll rewrite it later to avoid that.

**Review by kd-11 — APPROVED** [review]

**Pull request merged**